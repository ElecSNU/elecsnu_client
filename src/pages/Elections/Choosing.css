#choosing {
    height: 100%;
    padding: 5%;
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;
}

#choosing > * {
    width: 100%;
}

.candidate {
    padding: 2.5%;
    margin: 2.5%;
    transition: all 0.7s;
    border: 5px solid transparent;
    cursor: pointer;
}

.candidate h1 {
    transition: all 0.7s;
}

.candidate:hover,
.candidate.chosen {
    border: 5px solid var(--accent1);
}

.candidate:hover h1,
.candidate.chosen h1 {
    color: var(--accent1);
}

#choosing .submit-button {
    width: 50%;
}
