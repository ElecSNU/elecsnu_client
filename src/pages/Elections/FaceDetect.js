import React, { useEffect, useRef } from 'react';
import Webcam from 'react-webcam';
import * as faceapi from 'face-api.js';
import { fire_storage } from '../../configs/firebase/config';
import { b64toBlob } from './HelperFunctions';

import './FaceDetect.css';

const FaceDetect = () => {
    const videoRef = useRef();
    const webcamRef = useRef(null);

    const playVideo = async () => {
        await loadFaceDetectModels();
        navigator.getUserMedia(
            { video: {} },
            (stream) =>
                (videoRef.current.srcObject = stream),
            (e) => console.error(e)
        );
    };

    const loadFaceDetectModels = async () => {
        await faceapi.nets.tinyFaceDetector
            .loadFromUri('/models')
            .then((result) =>
                console.log('SUCCESS', result)
            )
            .catch((e) => console.log('ERROR', e));
    };

    const faceDetect = async () => {
        // await loadFaceDetectModels();
        const detections2 = await faceapi.detectAllFaces(
            videoRef.current,
            new faceapi.TinyFaceDetectorOptions()
        );
        if (detections2.length === 1) {
            if (capture())
                setTimeout(() => {
                    const elecId = window.location.pathname.split(
                        '/'
                    )[3];

                    window.location.pathname = `/dashboard/poll/${elecId}/voting/detected`;
                }, 500);

            videoRef.current.pause();
        }
    };

    const capture = () => {
        let screenshot = null;
        if (webcamRef && webcamRef.current)
            screenshot = webcamRef.current.getScreenshot();

        let ssBlob = null;
        if (screenshot)
            ssBlob = b64toBlob(
                screenshot.split(',')[1],
                'image/jpeg'
            );

        console.log('HEY');
        if (ssBlob) {
            // uploadImage(ssBlob, fire_storage);
            fire_storage
                .ref('face_image')
                .put(ssBlob)
                .then(() => {
                    window.localStorage.setItem(
                        'imageUploaded',
                        'y'
                    );
                });
            // console.log(imageLink);
            return true;
        }
    };

    useEffect(() => {
        window.localStorage.removeItem('imageUploaded');
        // loadFaceDetectModels();
        playVideo();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    // useEffect(() => {
    //     if (faceDetectStatus) showLoader(true);
    // }, [faceDetectStatus]);

    return (
        <div id='face-detect'>
            <h1 className='heading-text foreground-back center-text capital text'>
                Verifying voter Identity
            </h1>
            <br />
            {/* {faceDetectStatus ? (
                <></>
            ) : ( */}
            <>
                <video
                    ref={videoRef}
                    autoPlay
                    style={{
                        transform: 'scale(-1, 1)',
                    }}
                    className='border-radius-15 box-shadow video'
                    onPlay={() =>
                        setInterval(() => faceDetect(), 500)
                    }
                ></video>
                <Webcam
                    audio={false}
                    height={720}
                    ref={webcamRef}
                    screenshotFormat='image/jpeg'
                    width={1280}
                    className='video'
                    style={{
                        opacity: '0',
                        position: 'absolute',
                    }}
                />
            </>
            {/* )}   */}
        </div>
    );
};

export default FaceDetect;
