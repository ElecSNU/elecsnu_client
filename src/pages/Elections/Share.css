#share {
    width: 100vw;
    height: 100vh;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    background-color: rgba(0, 0, 0, 0.7);
    display: grid;
    place-items: center;
    animation: show 0.7s ease-in-out;
}

@keyframes show {
    0% {
        opacity: 0;
    }

    100% {
        opacity: 1;
    }
}

.share-modal {
    position: relative;
    width: 75%;
}

.cross {
    font-size: 2rem;
    color: red;
    position: absolute;
    top: 10%;
    right: 2.5%;
    cursor: pointer;
}
