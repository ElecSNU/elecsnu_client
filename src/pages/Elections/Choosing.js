import React, { useState } from 'react';
import useFirestore from '../../hooks/useFirestore';
import { fire_store } from '../../configs/firebase/config';
import { toast } from 'react-toastify';
import useLoader from '../../hooks/useLoader';

import Share from './Share';

import './Choosing.css';

const Choosing = ({ candidates }) => {
    const [chosen, setChosen] = useState(0);
    const electionDocs = useFirestore('newelections').docs;
    const showLoader = useLoader();
    const [done, setDone] = useState(false);
    const [electionName, setElectionName] = useState('');

    const castVote = () => {
        showLoader(true);
        const electionId = window.location.pathname.split(
            '/'
        )[3];
        let electionDetails = {};
        electionDocs.forEach((e) => {
            if (e.id === electionId) {
                electionDetails = e;
                return;
            }
        });
        setElectionName(electionDetails.election_name);

        let userGender = window.localStorage.getItem(
            'voter_gender'
        );
        let userDept = window.localStorage.getItem(
            'voter_dept'
        );
        let userBatch = window.localStorage.getItem(
            'voter_batch'
        );
        let userRoll = window.localStorage.getItem(
            'user_roll'
        );

        electionDetails.voters = [
            ...electionDetails.voters,
            Number(userRoll),
        ];
        electionDetails.election_votes = [
            ...electionDetails.election_votes,
            {
                voter_gender: userGender,
                groups: `${userDept},Batch${userBatch}`,
                candidate_chosen: chosen,
            },
        ];

        fire_store
            .collection('newelections')
            .doc(electionId)
            .set(electionDetails);

        showLoader(false);

        toast.success('Successfully casted vote!');

        setDone(true);
    };

    return (
        <>
            {done ? (
                <Share electionName={electionName} />
            ) : (
                <></>
            )}
            <div id='choosing'>
                <div className='heading-text capital-text center-text foreground-back'>
                    Who would you want to vote for?
                </div>
                <div className='candidates'>
                    {candidates.map((candidate, index) => {
                        return (
                            <div
                                className={`candidate background-1 border-radius-15 ${
                                    index === chosen
                                        ? 'chosen'
                                        : ''
                                }`}
                                key={index}
                                onClick={() =>
                                    setChosen(index)
                                }
                            >
                                <h1 className='capital-text subheading-text foreground-light center-text'>
                                    {candidate}
                                </h1>
                            </div>
                        );
                    })}
                </div>
                <button
                    className='submit-button'
                    onClick={castVote}
                >
                    CAST VOTE
                </button>
            </div>
        </>
    );
};

export default Choosing;
