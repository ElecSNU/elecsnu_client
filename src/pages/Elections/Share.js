import React from 'react';

import './Share.css';

// import icons
import FaTwitter from 'react-icons/lib/fa/twitter';
import FaFacebook from 'react-icons/lib/fa/facebook';
import FaEnvelope from 'react-icons/lib/fa/envelope';
import FaLinkedin from 'react-icons/lib/fa/linkedin';

// import react-custom-share components
import {
    ShareButtonRectangle,
    ShareBlockStandard,
} from 'react-custom-share';

const ShareComponent = (props) => {
    // create object with props for shareBlock
    const shareBlockProps = {
        url: 'elecsnu.firebaseapp.com',
        button: ShareButtonRectangle,
        buttons: [
            { network: 'Twitter', icon: FaTwitter },
            { network: 'Facebook', icon: FaFacebook },
            { network: 'Email', icon: FaEnvelope },
            { network: 'Linkedin', icon: FaLinkedin },
        ],
        text: 'Voting using ElecSNU',
        longtext: `Hey! I just voted for ${props.name} using ElecSNU. Check out their website for more!`,
    };

    return <ShareBlockStandard {...shareBlockProps} />;
};

const Share = ({ electionName }) => {
    return (
        <div id='share'>
            <div className='share-modal'>
                <h1
                    className='heading-text center-text foreground-accent1 border-radius-15 background-1'
                    style={{
                        padding: '2.5%',
                    }}
                >
                    Share your voting experience with
                    everyone!
                </h1>
                <div
                    className='cross'
                    onClick={() => {
                        window.location.pathname = '/';
                    }}
                >
                    x
                </div>
                <ShareComponent name={electionName} />
            </div>
        </div>
    );
};

export default Share;
