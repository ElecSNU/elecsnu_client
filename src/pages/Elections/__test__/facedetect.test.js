import { render, screen, fireEvent, cleanup } from '@testing-library/react';
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Choosing from '../Share'
import { toast } from 'react-toastify';
import React, { useEffect, useState } from 'react';
import Share from '../FaceDetect';
import FaceDetect from '../FaceDetect';
// import Elections from './../Elections';
// import ParticleBackground from '../../components/ParticleBackground/ParticleBackground';


test('Tests the Sharing ability', () => {
    const div = document.createElement("div")
    const tree = renderer.create(<FaceDetect></FaceDetect>).toJSON();
    expect(tree).toMatchSnapshot(); 
   });