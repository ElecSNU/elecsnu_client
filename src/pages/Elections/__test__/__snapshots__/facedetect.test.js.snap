// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests the Sharing ability 1`] = `
<div
  id="face-detect"
>
  <h1
    className="heading-text foreground-back center-text capital text"
  >
    Verifying voter Identity
  </h1>
  <br />
  <video
    autoPlay={true}
    className="border-radius-15 box-shadow video"
    onPlay={[Function]}
    style={
      Object {
        "transform": "scale(-1, 1)",
      }
    }
  />
  <video
    autoPlay={true}
    className="video"
    height={720}
    muted={false}
    playsInline={true}
    style={
      Object {
        "opacity": "0",
        "position": "absolute",
      }
    }
    width={1280}
  />
</div>
`;
